[Lockdown Mode] Per app mechanism to opt out of first-time-use prompt

brentfulgham opened the pull request:
#### ec6bbaa54fd8c0b5c94e26fa650c68ac93db8a8c
<pre>
[Lockdown Mode] Per app mechanism to opt out of first-time-use prompt
rdar://92423822

Reviewed by Chris Dumez.

Create a mechanism for apps to control the first-time-use behavior in Lockdown Mode.

* Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
(+[WKProcessPool _captivePortalModeEnabledGloballyForTesting]):
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPoolPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _initializeWithConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(isLockdownModeWarningNeeded):
(-[WKWebView _presentCaptivePortalMode]):
(-[WKWebView _presentCaptivePortalModeAlertIfNeeded]):
(+[WKWebView _resetPresentLockdownModeMessage]):
(-[WKWebView _presentCaptivePortalModeAlertOnce]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
(-[NoLockdownFirstUseMessage webView:showLockdownModeFirstUseMessage:completionHandler:]):
(-[AskAgainFirstUseMessage webView:showLockdownModeFirstUseMessage:completionHandler:]):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView setUpAndAddToTestWindow]):

Canonical link: <a href="https://commits.webkit.org/252700@main">https://commits.webkit.org/252700@main</a>
</pre>

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
```diff
@@ -505,6 +505,11 @@ + (void)_setCaptivePortalModeEnabledGloballyForTesting:(BOOL)isEnabled
     WebKit::setCaptivePortalModeEnabledGloballyForTesting(!!isEnabled);
 }
 
++ (BOOL)_captivePortalModeEnabledGloballyForTesting
```
Shouldn't we use the lockdown mode naming in new code at least? I think it's a matter of time until we rename every "captivePortalMode" to "lockdownMode" in existing code, no?

brentfulgham replied:
OK -- Will fix.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKUIDelegate.h:
```diff
@@ -57,6 +57,18 @@ typedef NS_ENUM(NSInteger, WKMediaCaptureType) {
+/*! @enum WKDialogResult
+@abstract Constants returned by showLockdownModeFirstUseMessage to indicate how WebKit should treat first use.
+@constant WKDialogResultShowDefault Indicates that the client did not display a first use message. WebKit should show the default.
+@constant WKDialogResultAskAgain Indicates the client handled the message, but want's to be checked if other WKWebViews are used.
```
Typo: "want's"

brentfulgham replied:
Doh! Fixed.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKUIDelegate.h:
```diff
@@ -57,6 +57,18 @@ typedef NS_ENUM(NSInteger, WKMediaCaptureType) {
+@constant WKDialogResultAskAgain Indicates the client handled the message, but want's to be checked if other WKWebViews are used.
+@constant WKDialogResultHandled Indicates the client handled the message and no further checks are needed.
+*/
+typedef NS_ENUM(NSInteger, WKDialogResult) {
```
Based on the documentation, this enum seems very specific to lock down mode's first use dialog. Based on that, using a generic name such as WKDialogResult is a bit confusing to me. Why aren't we using a more specific type name?

brentfulgham replied:
This was feedback that James gave during the API review process. I'd prefer to stick with this naming, since it's in the official API review submission, until a final approval is made. Feel free to pile on the API review discussion if you want the naming to be reconsidered (I'll e-mail you a link).

cdumez commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3002,44 +3003,94 @@ - (void)_enhancedWindowingToggled:(NSNotification *)notification
+    auto message = WEB_UI_NSSTRING(@"Certain experiences and features may not function as expected. You can turn off Lockdown Mode for this app in Settings.", "Lockdown Mode alert message");
 
-    [alert addAction:[UIAlertAction actionWithTitle:WEB_UI_NSSTRING(@"OK", "Captive Portal Mode alert OK button") style:UIAlertActionStyleDefault handler:nil]];
+    auto decisionHandler = makeBlockPtr([appDisplayName, message, self](WKDialogResult result) mutable {
```
Seems dangerous to not capture a retainSelf? Or maybe we don't need to capture self at all?

brentfulgham replied:
We do need self (at least it's a compile error without it), so I'll capture a retainSelf.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -389,6 +391,135 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+    [WKWebView _resetPresentLockdownModeMessage];
+
+    showedNoFirstUseMessage = false;
+    showedCustomFirstUseMessage = false;
```
This test doesn't not use this at all so why set it?

brentfulgham replied:
Fixed.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -389,6 +391,135 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+
+    showedNoFirstUseMessage = false;
+    showedCustomFirstUseMessage = false;
+    requestFutureFirstUseMessage = false;
```
ditto.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -389,6 +391,135 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+    TestWebKitAPI::Util::run(&done);
+
+    EXPECT_TRUE(showedNoFirstUseMessage);
+    EXPECT_FALSE(showedCustomFirstUseMessage);
```
This test could never set it. Under no condition could this global ever be set by this test so why check it?

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -389,6 +391,135 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+
+    EXPECT_TRUE(showedNoFirstUseMessage);
+    EXPECT_FALSE(showedCustomFirstUseMessage);
+    EXPECT_FALSE(requestFutureFirstUseMessage);
```
ditto.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -389,6 +391,135 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+
+    TestWebKitAPI::Util::run(&done);
+
+    EXPECT_TRUE(showedNoFirstUseMessage);
```
Overall, I am not sure what this test achieves. It just shows that the showLockdownModeFirstUseMessage delegate was indeed called. It doesn't really validate that no first use message is shown. WebKit could still be showing its own generic alert for e.g.

brentfulgham replied:
Do you know of a way to confirm that an unrelated alert UI is not displayed on screen? I couldn't find an example of anything that wasn't a developer-controlled UI where we could set state ourselves.

cdumez replied:
Maybe by swizzling`[UIAlertController alertControllerWithTitle:` or `[presentationViewController presentViewController:]` and setting a global boolean in your swizzled implementation. You'd be able to observe if WebKit tried to show the later or not.

brentfulgham replied:
Ah! Gotcha. Let me try that.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -389,6 +391,135 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+    [WKProcessPool _setCaptivePortalModeEnabledGloballyForTesting:YES];
+    [WKWebView _resetPresentLockdownModeMessage];
+
+    showedNoFirstUseMessage = false;
```
Not used by this test.

brentfulgham replied:
Fixed!

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -389,6 +391,135 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+
+    TestWebKitAPI::Util::run(&done);
+
+    EXPECT_FALSE(showedNoFirstUseMessage);
```
Why are we checking this? This could never be set by this test.

brentfulgham replied:
Fixed.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -389,6 +391,135 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+
+    TestWebKitAPI::Util::run(&done);
+
+    EXPECT_FALSE(showedNoFirstUseMessage);
```
Why are we checking this? This could never be set by this test.

brentfulgham replied:
Fixed.

cdumez commented on Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
```diff
@@ -524,6 +524,13 @@ - (void)_setUpTestWindow:(NSRect)frame
+{
+    ASSERT(!_hostWindow);
+    [self _setUpTestWindow:self.frame];
+    [self addToTestWindow];
```
Seems _setUpTestWindow already does what addToTestWindow does.

Also, why do we need this new function at all. Why not pass `addToWindow:YES` when constructing the TestWKWebView? (which would call the internal `_setUpTestWindow` for you).

brentfulgham replied:
If we use 'addToWindow:YES' at construction time, it creates the window and adds the view to it before we have a chance to set the UI delegate. This causes the first-use logic to fire before we have a chance to register our own handler.

cdumez replied:
I see. Then maybe we could have the tests call `[webView addToTestWindow]` and update `addToTestWindow` to call `_setUpTestWindow` internally if `_hostWindow` is nil.

brentfulgham replied:
I have a better approach that will be less confusing. I'll update the patch.

brentfulgham commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -389,6 +390,113 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+constexpr auto WebKitCaptivePortalModeAlertShownKey = @"WebKitCaptivePortalModeAlertShown";
+
+static bool showedNoFirstUseMessage;
+
```
Whitespace fail. :-(

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
```diff
@@ -156,6 +156,8 @@ enum class TapHandlingResult : uint8_t;
 - (void)_beginLiveResize;
 - (void)_endLiveResize;
 
++ (void)_resetPresentLockdownModeMessage;
```
This could be moved to `WKWebViewPrivateForTestingIOS.h`.

brentfulgham replied:
Sounds good. It is actually backed by a static 'bool' in the `WKWebViewIOS.mm` file, but I can expose it through the `WKWebViewPrivateForTestingIOS.h `interface.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -389,6 +390,112 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+#if PLATFORM(IOS)
+@interface WKWebView (WKViewInternalIOS)
++ (void)_resetPresentLockdownModeMessage;
+@end
```
(and then you can remove this declaration)

brentfulgham replied:
Done.

cdumez commented on Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
```diff
@@ -85,6 +86,7 @@
 @interface TestWKWebView : WKWebView
 - (instancetype)initWithFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration processPoolConfiguration:(_WKProcessPoolConfiguration *)processPoolConfiguration;
 - (instancetype)initWithFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration addToWindow:(BOOL)addToWindow;
+- (instancetype)initWithFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration uiDelegate:(id<WKUIDelegate>)delegate addToWindow:(BOOL)addToWindow;
```
For what it's worth, I'm not a fan of the fix you went with.

You could have fixed this by:
1. Using a regular WKWebView
2. Using a testWKWebView, call addToWindow after you've added your UIDelegate and make sure that the addToWindow implementation constructs the hostWindow if necessary.

Instead we have now yet another TestWKWebView init function, which is unlikely to be reused since callers can set the WKUIDelegate themselves.

brentfulgham replied:
I'll switch back to the original state and adopt this approach.

cdumez commented on Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
```diff
@@ -526,6 +527,9 @@ - (void)_setUpTestWindow:(NSRect)frame
 - (void)addToTestWindow
 {
+    if (!_hostWindow)
+        [self _setUpTestWindow:self.frame];
```
We can return after this. _setUpTestWindow already runs the logic below.

brentfulgham replied:
Ah! Good catch!

cdumez reviewed (approved):
LGTM

brentfulgham committed:
[Lockdown Mode] Per app mechanism to opt out of first-time-use prompt
rdar://92423822

Reviewed by Chris Dumez.

Create a mechanism for apps to control the first-time-use behavior in Lockdown Mode.

* Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
(+[WKProcessPool _captivePortalModeEnabledGloballyForTesting]):
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPoolPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _initializeWithConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(isLockdownModeWarningNeeded):
(-[WKWebView _presentCaptivePortalMode]):
(-[WKWebView _presentCaptivePortalModeAlertIfNeeded]):
(+[WKWebView _resetPresentLockdownModeMessage]):
(-[WKWebView _presentCaptivePortalModeAlertOnce]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
(-[NoLockdownFirstUseMessage webView:showLockdownModeFirstUseMessage:completionHandler:]):
(-[AskAgainFirstUseMessage webView:showLockdownModeFirstUseMessage:completionHandler:]):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView setUpAndAddToTestWindow]):

Canonical link: https://commits.webkit.org/252700@main
